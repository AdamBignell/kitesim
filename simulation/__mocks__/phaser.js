const createMockSceneSystems = () => ({
  add: {
    graphics: jest.fn(() => ({
      fillStyle: jest.fn().mockReturnThis(),
      fillRect: jest.fn().mockReturnThis(),
      generateTexture: jest.fn().mockReturnThis(),
      destroy: jest.fn(),
    })),
    rectangle: jest.fn(() => ({
      setOrigin: jest.fn().mockReturnThis(),
    })),
    text: jest.fn(),
  },
  make: {
    tilemap: jest.fn(() => ({
      addTilesetImage: jest.fn(),
      createLayer: jest.fn(() => ({ setCollision: jest.fn(), destroy: jest.fn() })),
      setCollision: jest.fn(),
    })),
  },
  physics: {
    add: {
      staticGroup: jest.fn(() => ({
        add: jest.fn(),
        clear: jest.fn(),
        create: jest.fn(() => ({
          setSize: jest.fn().mockReturnThis(),
          setVisible: jest.fn().mockReturnThis(),
          refreshBody: jest.fn().mockReturnThis(),
        })),
        refresh: jest.fn(),
      })),
      sprite: jest.fn(() => ({
        setCircle: jest.fn(),
        setBounce: jest.fn(),
        setCollideWorldBounds: jest.fn(),
        setPosition: jest.fn(),
        setVelocity: jest.fn(),
        setDepth: jest.fn(),
        body: {
            getBounds: jest.fn(() => ({ bottom: 0})),
            touching: {},
        }
      })),
      collider: jest.fn(),
    },
    world: {
      setBounds: jest.fn(),
      gravity: { y: 1500 },
    },
  },
  cameras: {
    main: {
      setBackgroundColor: jest.fn(),
    },
  },
  scale: {
    on: jest.fn(),
    width: 1280,
    height: 720,
    gameSize: { width: 1280, height: 720 },
  },
  input: {
    keyboard: {
      createCursorKeys: jest.fn(() => ({})),
      addKey: jest.fn(),
      addKeys: jest.fn(),
      on: jest.fn(),
    },
  },
  time: {
    addEvent: jest.fn(),
    delayedCall: jest.fn(),
  },
  anims: {
    create: jest.fn(),
    generateFrameNumbers: jest.fn(),
  }
});

const Scene = jest.fn().mockImplementation(function() {
  // Assign a fresh set of mocks for each scene instance
  Object.assign(this, createMockSceneSystems());
});

module.exports = {
  Scene: Scene,
  Scale: { FIT: 'FIT', CENTER_BOTH: 'CENTER_BOTH' },
  AUTO: 'AUTO',
  Input: { Keyboard: { KeyCodes: { SHIFT: 'SHIFT' }, JustDown: jest.fn() } },
  Math: {
    Between: jest.fn((min) => min),
    Clamp: jest.fn((value, min, max) => Math.max(min, Math.min(value, max))),
    Vector2: jest.fn().mockImplementation(function (x, y) {
        this.x = x || 0;
        this.y = y || 0;
        this.clone = () => new module.exports.Math.Vector2(this.x, this.y);
        this.scale = jest.fn().mockReturnThis();
    }),
    Distance: { BetweenPoints: jest.fn(() => 100) }
  },
  Geom: { Line: jest.fn(() => ({ getPoints: jest.fn(() => []) })) }
};
