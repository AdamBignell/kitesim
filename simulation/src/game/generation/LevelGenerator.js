import * as Phaser from 'phaser';
import Room from './Room';
import structures from './structures';
import TraversalValidator from './TraversalValidator';

export default class LevelGenerator {
  constructor(scene) {
    this.scene = scene;
  }

  generate(platformsGroup) {
    const { width, height } = this.scene.scale;
    const wallThickness = 20;
    const openingSize = 100;

    // --- Create Outer Walls with Openings ---

    // Floor (with opening)
    const floorLeftWidth = width / 2 - openingSize / 2;
    const floorRightWidth = width / 2 - openingSize / 2;
    const floorLeft = this.scene.add.rectangle(floorLeftWidth / 2, height - wallThickness / 2, floorLeftWidth, wallThickness, 0x000000);
    const floorRight = this.scene.add.rectangle(width - floorRightWidth / 2, height - wallThickness / 2, floorRightWidth, wallThickness, 0x000000);
    platformsGroup.add(floorLeft);
    platformsGroup.add(floorRight);

    // Ceiling (with opening)
    const ceilLeftWidth = width / 2 - openingSize / 2;
    const ceilRightWidth = width / 2 - openingSize / 2;
    const ceilLeft = this.scene.add.rectangle(ceilLeftWidth / 2, wallThickness / 2, ceilLeftWidth, wallThickness, 0x000000);
    const ceilRight = this.scene.add.rectangle(width - ceilRightWidth / 2, wallThickness / 2, ceilRightWidth, wallThickness, 0x000000);
    platformsGroup.add(ceilLeft);
    platformsGroup.add(ceilRight);

    // Left Wall (with opening)
    const leftWallTopHeight = height / 2 - openingSize / 2;
    const leftWallBottomHeight = height / 2 - openingSize / 2;
    const leftWallTop = this.scene.add.rectangle(wallThickness / 2, leftWallTopHeight / 2, wallThickness, leftWallTopHeight, 0x000000);
    const leftWallBottom = this.scene.add.rectangle(wallThickness / 2, height - leftWallBottomHeight / 2, wallThickness, leftWallBottomHeight, 0x000000);
    platformsGroup.add(leftWallTop);
    platformsGroup.add(leftWallBottom);

    // Right Wall (with opening)
    const rightWallTopHeight = height / 2 - openingSize / 2;
    const rightWallBottomHeight = height / 2 - openingSize / 2;
    const rightWallTop = this.scene.add.rectangle(width - wallThickness /2, rightWallTopHeight / 2, wallThickness, rightWallTopHeight, 0x000000);
    const rightWallBottom = this.scene.add.rectangle(width - wallThickness / 2, height - rightWallBottomHeight / 2, wallThickness, rightWallBottomHeight, 0x000000);
    platformsGroup.add(rightWallTop);
    platformsGroup.add(rightWallBottom);

    // Generate a traversable room
    let room;
    let isValid = false;
    while (!isValid) {
      room = new Room(20, 15, structures);
      room.generateLayout();
      const validator = new TraversalValidator(room);
      isValid = validator.isTraversable();
    }

    room.render(this.scene, platformsGroup);

    platformsGroup.refresh();
  }
}
