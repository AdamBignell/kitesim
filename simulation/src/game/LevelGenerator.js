import * as Phaser from 'phaser';
import { createNoise2D } from 'simplex-noise';
// Grid and GreedyMesher are no longer needed
// import Grid from './generation/Grid';
// import GreedyMesher from './generation/GreedyMesher';
import * as Structures from './generation/structures';
import { createFloor } from './generation/MegaStructure';
import PlayerCapabilitiesProfile from './generation/PlayerCapabilitiesProfile';
import SplinePathGenerator from './generation/SplinePathGenerator';
import RhythmNodeCalculator from './generation/RhythmNodeCalculator';

export default class LevelGenerator {
  constructor(scene, pcp) {
    this.scene = scene;
    this.structures = Object.values(Structures);
    this.pcp = pcp;
    this.noise = createNoise2D();
    const rhythmCalculator = new RhythmNodeCalculator(pcp);
    this.pathGenerator = new SplinePathGenerator(rhythmCalculator);
  }

  // --- NEW generateChunk ---
  generateChunk(chunkX, chunkY, chunkSize, tileSize) {
    const chunkWorldX = chunkX * chunkSize * tileSize;
    // Using scene height for vertical bounds is more robust for a scrolling world
    const chunkWorldY = 0; // Top of the world
    const chunkHeightPixels = this.scene.scale.height * 2; // Make it tall enough

    const chunkWidthPixels = chunkSize * tileSize;


    // --- Vertex Generation ---
    const surfaceVertices = [];
    const terrainNoiseScale = 50;
    const terrainAmplitude = 15 * tileSize; // Amplitude in pixels
    const worldCenterY = this.scene.scale.height / 1.5; // Base height in pixels
    const step = 32; // Generate a vertex every 32 pixels (one tile)

    for (let x = 0; x <= chunkWidthPixels; x += step) {
      const worldX = chunkWorldX + x;

      const splinePoint = this.pathGenerator.getPointAtWorldX(worldX / 20);
      const splineHeight = splinePoint.y * tileSize / 10; // Adjust spline scale

      const noiseValue = this.noise(worldX / terrainNoiseScale, 0);
      const noiseOffset = noiseValue * terrainAmplitude;

      const terrainHeight = worldCenterY + splineHeight + noiseOffset;
      surfaceVertices.push({ x: worldX, y: terrainHeight });
    }

    // --- Create Polygon Shape ---
    const groundPolygon = [
      ...surfaceVertices,
      { x: chunkWorldX + chunkWidthPixels, y: chunkWorldY + chunkHeightPixels },
      { x: chunkWorldX, y: chunkWorldY + chunkHeightPixels }
    ];

    // --- Create Matter Body ---
    const terrainBody = this.scene.matter.add.fromVertices(0, 0, groundPolygon, { isStatic: true });


    // --- Visuals ---
    // The container will hold the graphics and a reference to the Matter body
    // This makes cleanup easier when the chunk is unloaded.
    const graphics = this.scene.add.graphics();
    graphics.fillStyle(0x228B22, 1); // ForestGreen
    graphics.fillPoints(groundPolygon);

    const container = this.scene.add.container(0, 0, [graphics]);
    container.setData('matterBody', terrainBody);
    // Override the container's destroy method to also remove the Matter body
    container.destroy = function() {
        graphics.destroy();
        // 'this' refers to the container instance
        this.scene.matter.world.remove(this.getData('matterBody'));
        // Call original destroy method
        Phaser.GameObjects.Container.prototype.destroy.call(this);
    };


    return { platforms: container };
  }


  // --- NEW generateInitialChunkAndSpawnPoint ---
  generateInitialChunkAndSpawnPoint(chunkSize, tileSize) {
    const chunkX = 0;
    const chunkWidthPixels = chunkSize * tileSize;
    const chunkHeightPixels = this.scene.scale.height * 2;

    // --- Vertex Generation ---
    const surfaceVertices = [];
    const terrainNoiseScale = 50;
    const terrainAmplitude = 15 * tileSize;
    const worldCenterY = this.scene.scale.height / 1.5;
    const step = 32;

    for (let x = 0; x <= chunkWidthPixels; x += step) {
      const worldX = x; // For initial chunk, worldX is same as local x

      const splinePoint = this.pathGenerator.getPointAtWorldX(worldX / 20);
      const splineHeight = splinePoint.y * tileSize / 10;

      const noiseValue = this.noise(worldX / terrainNoiseScale, 0);
      const noiseOffset = noiseValue * terrainAmplitude;

      const terrainHeight = worldCenterY + splineHeight + noiseOffset;
      surfaceVertices.push({ x: worldX, y: terrainHeight });
    }

    // --- Find Spawn Point ---
    const safeZoneTileX = 3; // ~3rd tile in
    const searchWorldX = safeZoneTileX * tileSize;
    let groundY = worldCenterY; // fallback

    // Find the terrain height at our desired spawn X by finding the closest vertex
    let closestVertex = surfaceVertices[0];
    let minDistance = Infinity;
    for(const vertex of surfaceVertices) {
        const distance = Math.abs(vertex.x - searchWorldX);
        if (distance < minDistance) {
            minDistance = distance;
            closestVertex = vertex;
        }
    }
    groundY = closestVertex.y;

    const spawnPoint = {
        x: searchWorldX,
        y: groundY - (tileSize * 3) // Spawn 3 tiles above the ground
    };


    // --- Create Polygon Shape ---
    const groundPolygon = [
      ...surfaceVertices,
      { x: chunkWidthPixels, y: chunkHeightPixels },
      { x: 0, y: chunkHeightPixels }
    ];

    // --- Create Matter Body ---
    const terrainBody = this.scene.matter.add.fromVertices(0, 0, groundPolygon, { isStatic: true });


    // --- Visuals ---
    const graphics = this.scene.add.graphics();
    graphics.fillStyle(0x228B22, 1); // ForestGreen
    graphics.fillPoints(groundPolygon);

    const container = this.scene.add.container(0, 0, [graphics]);
    container.setData('matterBody', terrainBody);
    container.destroy = function() {
        graphics.destroy();
        this.scene.matter.world.remove(this.getData('matterBody'));
        Phaser.GameObjects.Container.prototype.destroy.call(this);
    };

    return { platforms: container, spawnPoint };
  }
}
