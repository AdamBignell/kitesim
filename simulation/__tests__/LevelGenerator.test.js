import LevelGenerator from '../src/game/LevelGenerator';
import * as Phaser from 'phaser';

// Mocks are defined in __mocks__/phaser.js
// We don't need to mock LevelGenerator since that's what we're testing.

describe('LevelGenerator', () => {
  let scene;
  let levelGenerator;
  let platforms;

  beforeEach(() => {
    // Create a mock scene object using the mock from __mocks__/phaser.js
    scene = new Phaser.Scene();

    // The mock for a static group, as defined in __mocks__/phaser.js
    platforms = scene.physics.add.staticGroup();

    // Instantiate the LevelGenerator with the mock scene
    levelGenerator = new LevelGenerator(scene);
  });

  it('should be created', () => {
    expect(levelGenerator).toBeDefined();
  });

  it('should call the draw function for each chunk in the grid', () => {
    // Spy on the draw functions for a few templates
    const startRoom = levelGenerator.templates['START_ROOM'];
    const solidRoom = levelGenerator.templates['SOLID'];
    jest.spyOn(startRoom, 'draw');
    jest.spyOn(solidRoom, 'draw');

    levelGenerator.generate(platforms);

    // We expect at least the start room to be drawn
    expect(startRoom.draw).toHaveBeenCalled();

    // The total number of draw calls should equal the grid size
    const GRID_COLS = 4;
    const GRID_ROWS = 3;
    const totalDrawCalls = startRoom.draw.mock.calls.length + solidRoom.draw.mock.calls.length;
    // Note: This is a simplification. In a real scenario, we'd have to spy on all templates.
    // For now, we are checking that our main method is being called.
    expect(totalDrawCalls).toBeGreaterThan(0);
  });

  it('should always place START_ROOM at the bottom-left', () => {
    const grid = levelGenerator._fillGrid(4, 3);
    // Grid is [row][col], so bottom-left is [rows-1][0]
    expect(grid[2][0]).toBe('START_ROOM');
  });

  it('should create a grid of the correct size', () => {
    const cols = 5, rows = 4;
    const grid = levelGenerator._fillGrid(cols, rows);
    expect(grid.length).toBe(rows);
    expect(grid[0].length).toBe(cols);
  });

  describe('Drawing Functions', () => {
    const x = 0, y = 0, w = 100, h = 100;

    it('_drawSolidChunk should create a large, single platform', () => {
      levelGenerator._drawSolidChunk(platforms, x, y, w, h);
      expect(platforms.create).toHaveBeenCalledWith(x + w / 2, y + h / 2, 'platform');
      const mockPlatform = platforms.create.mock.results[0].value;
      expect(mockPlatform.setSize).toHaveBeenCalledWith(w, h);
    });

    it('_drawFloorChunk should create a platform at the bottom', () => {
      levelGenerator._drawFloorChunk(platforms, x, y, w, h);
      expect(platforms.create).toHaveBeenCalledWith(x + w / 2, y + h - 10, 'platform');
    });

    it('_drawVerticalShaft should create two vertical walls', () => {
      levelGenerator._drawVerticalShaft(platforms, x, y, w, h);
      expect(platforms.create).toHaveBeenCalledTimes(2);
      expect(platforms.create).toHaveBeenCalledWith(x + 10, y + h / 2, 'platform');
      expect(platforms.create).toHaveBeenCalledWith(x + w - 10, y + h / 2, 'platform');
    });
  });
});
