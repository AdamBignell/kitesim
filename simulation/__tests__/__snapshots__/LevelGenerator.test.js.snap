// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LevelGenerator should generate a chunk that matches the snapshot 1`] = `
"................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
................................
...........##############.......
...........################.....
...........################.....
...........################.....
...........################.....
...........################.....
...........################.....
...........################.....
...........################.....
...........################.....
...........################.....
...........################.....
"
`;
